### PR TITLE
Add `isExpr`, `asExpr`, `asExprOf`, `asTerm`, `asTypeTree` and `asQuotedType`

### PR DESCRIPTION
**Pull request by nicolasstucki**
These replace the `Expr` methods `seal`, `sealOpt` and `unseal` with equivalent methods with more intuitive names.
Instead of `sealOpt` we provide a more flexible `isExpr`. All these methods are defined directly on `Tree`
instead of `Term` to avoid the need to perform type test to a `Term` first.

We similarly replace `quoted.Type.seal` with `asTypeTree` and `reflect.Type.seal` with `asQuotedType`.

Note: #8940 would remove the `asTerm` and `asTypeTree` as there we have `Expr[T] <:< Term` and something similar for `Type`, but the others would still be relevant.

### CODE REVIEW

**Comment by liufengyun** [comment]
The names `seal` and `unseal` look good to me, meta-programmers get to know them better in usage.

That said, I'm open to changes, the only worry is about compatibility -- it seems the change is pervasive and will break most macros written for Dotty.

**Comment by nicolasstucki** [comment]
The old version will still be supported for some time with a deprecation warning

**Comment by nicolasstucki** [comment]
The names `seal` and `unseal` also look good to me because I have used them for years. But not to a newcomer that wants to convert an `Expr` to a `Term`, `Term` to an `Expr`, and so on ... The new names tell exactly what is the conversion and can be easily found by autocompletion.

The `toExprOf` is the most important one as it gives a way to get an `Expr[T]` from any `Expr` or `Tree` while checking that it is an expression and that it is in fact of type `T`. I saw too many instances of `term.seal.cast[T]` and `term.seal.asInstanceOf[Expr[T]]`; the second one is particularly dangerous as the it is an unchecked cast that leads to unsoundness that shows in `-Ycheck` in later phases.

**Comment by AdamPaynter** [comment]
I can confirm that, as a *very* new newcomer, this makes more sense. In fact, I hadn't even figured out what `seal` and `unseal` were until I read this PR. Even now, I still haven't grokked the relationship between `Term` and `Expr` (and why you would switch between them). Time for some more reading...

File: library/src-bootstrapped/scala/quoted/Type.scala [hunk 1]
```diff
@@ -9,14 +9,18 @@ abstract class Type[X <: AnyKind] private[scala] {
...
+    this.asTypeTree.showWith(SyntaxHighlight.plain)
 
   /** Show a source code like representation of this type */
   def showWith(syntaxHighlight: SyntaxHighlight)(using qctx: QuoteContext): String =
-    this.unseal.showWith(syntaxHighlight)
+    this.asTypeTree.showWith(syntaxHighlight)
 
   /** View this expression `quoted.Type[T]` as a `TypeTree` */
-  def unseal(using qctx: QuoteContext): qctx.tasty.TypeTree
+  @deprecated("Replaced with `asExprOf`", "0.27.0")
```
Comment (bishabosha) [inline library/src-bootstrapped/scala/quoted/Type.scala:19]:
```suggestion
  @deprecated("Replaced with `asTypeTree`", "0.27.0")
```

**Comment by nicolasstucki** [comment]
Replaced with #9514. Will do this in smaller parts

**Pull request closed without merging**